File: sdks/go/capability_manifest.go
package blok

import "encoding/json"

// CapabilityManifest is the language-neutral v1 operational contract from ADR 0003.
// Secret entries are opaque reference names, never credential values.
type CapabilityManifest struct {
	Version        string                    `json:"version"`
	Classification string                    `json:"classification"`
	Effects        []string                  `json:"effects"`
	Capabilities   []string                  `json:"capabilities"`
	Secrets        []string                  `json:"secrets"`
	Determinism    string                    `json:"determinism"`
	Idempotency    string                    `json:"idempotency"`
	Maturity       string                    `json:"maturity"`
	Resources      *CapabilityResourceBounds `json:"resources,omitempty"`
	Runtimes       []string                  `json:"runtimes,omitempty"`
	Triggers       []string                  `json:"triggers,omitempty"`
}

type CapabilityResourceBounds struct {
	MaxDurationMs  int64 `json:"maxDurationMs,omitempty"`
	MaxMemoryBytes int64 `json:"maxMemoryBytes,omitempty"`
	MaxInputBytes  int64 `json:"maxInputBytes,omitempty"`
	MaxOutputBytes int64 `json:"maxOutputBytes,omitempty"`
	MaxConcurrency int64 `json:"maxConcurrency,omitempty"`
}

// JSON returns the manifest's wire representation for gRPC ListNodes.
func (m CapabilityManifest) JSON() []byte {
	b, err := json.Marshal(m)
	if err != nil {
		return nil
	}
	return b
}

File: sdks/rust/src/capability_manifest.rs
use serde::{Deserialize, Serialize};

/// Language-neutral v1 operational metadata from ADR 0003.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CapabilityManifest {
    pub version: String,
    pub classification: String,
    pub effects: Vec<String>,
    pub capabilities: Vec<String>,
    pub secrets: Vec<String>,
    pub determinism: String,
    pub idempotency: String,
    pub maturity: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resources: Option<CapabilityResourceBounds>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub runtimes: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub triggers: Option<Vec<String>>,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CapabilityResourceBounds {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_duration_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_memory_bytes: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_input_bytes: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_output_bytes: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_concurrency: Option<u64>,
}
